test/e2e/testsuites/dynamically_provisioned_pod_node_selector.go

```diff
@@ -0,0 +1,129 @@
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	azDiskClientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+	consts "sigs.k8s.io/azuredisk-csi-driver/pkg/azureconstants"
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/azureutils"
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/controller"
+	"sigs.k8s.io/azuredisk-csi-driver/test/e2e/driver"
+)
+
+//  will provision required PV(s), PVC(s) and Pod(s)
+// Pod should successfully be re-scheduled on failover in a cluster with AzDriverNode and AzVolumeAttachment resources
+type PodNodeSelector struct {
+	CSIDriver              driver.DynamicPVTestDriver
+	Pod                    PodDetails
+	Volume                 VolumeDetails
+	AzDiskClient           *azDiskClientSet.Clientset
+	StorageClassParameters map[string]string
+}
+
+func (t *PodNodeSelector) Run(client clientset.Interface, namespace *v1.Namespace, schedulerName string) {
+	_, maxMountReplicaCount := azureutils.GetMaxSharesAndMaxMountReplicaCount(t.StorageClassParameters)
+
+	// Get the list of available nodes for scheduling the pod
+	nodes := ListNodeNames(client)
+	if len(nodes) < maxMountReplicaCount+1 {
+		ginkgo.Skip("need at least %d nodes to verify the test case. Current node count is %d", maxMountReplicaCount+1, len(nodes))
+	}
+
+	ctx := context.Background()
+
+	// set node label
+	numNodesWithLabel := maxMountReplicaCount + 1
+	nodesWithLabel := map[string]struct{}{}
+	for i := 0; i < numNodesWithLabel; i++ {
+		nodeObj, err := client.CoreV1().Nodes().Get(ctx, nodes[i], metav1.GetOptions{})
+		framework.ExpectNoError(err)
+		labelCleanup, err := SetNodeLabels(client, nodeObj, testLabel)
+		framework.ExpectNoError(err)
+		defer labelCleanup()
+		nodesWithLabel[nodes[i]] = struct{}{}
+	}
+
+	tpod, cleanup := t.Pod.SetupWithDynamicVolumes(client, namespace, t.CSIDriver, t.StorageClassParameters, schedulerName)
+	// defer must be called here for resources not get removed before using them
+	for i := range cleanup {
+		defer cleanup[i]()
+	}
+
+	tpod.SetNodeSelector(testLabel)
+	ginkgo.By("deploying the pod")
+	tpod.Create()
+	defer tpod.Cleanup()
+	ginkgo.By("checking that the pod is running")
+	tpod.WaitForRunning()
+	framework.ExpectNotEqual(t.Pod.Volumes[0].PersistentVolume, nil)
+
+	diskNames := make([]string, len(t.Pod.Volumes))
+	for i, volume := range t.Pod.Volumes {
+		framework.ExpectNotEqual(volume.PersistentVolume, nil)
+
+		pv, err := client.CoreV1().PersistentVolumes().Get(ctx, volume.PersistentVolume.Name, metav1.GetOptions{})
+		framework.ExpectNoError(err)
+		framework.ExpectNotEqual(pv.Spec.CSI, nil)
+		framework.ExpectEqual(pv.Spec.CSI.Driver, consts.DefaultDriverName)
+
+		diskName, err := azureutils.GetDiskName(pv.Spec.CSI.VolumeHandle)
+		framework.ExpectNoError(err)
+		diskNames[i] = strings.ToLower(diskName)
+	}
+
+	// confirm that primary and replica AzVolumeAttachment nodes for this volume is not created on a node without the taint
```

Comment:
np: double-negative makes this hard to parse and doesn't seem to be quite correct. Maybe "Confirm that the primary and replica AzVolumeAttachments for this volume are created on a node without the label"?